Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift

```diff
@@ -570,8 +615,200 @@ extension MessageHistoryEndpointTests {
+        case let .failure(error):
+          XCTFail("Message Counts request failed with error: \(error.localizedDescription)")
+        }
+        expectation.fulfill()
+      }
+
+    wait(for: [expectation], timeout: 1.0)
+  }
+
+  func testMessageCounts_Error_InvalidArguments() {
+    let expectation = self.expectation(description: "Message Counts Response Received")
+
+    guard let sessions = try? MockURLSession.mockSession(for: ["message_counts_error_invalid_arguments"]) else {
+      return XCTFail("Could not create mock url session")
+    }
+
+    PubNub(configuration: config, session: sessions.session)
+      .messageCounts(channels: testSingleChannel, timetoken: 0) { result in
+        switch result {
+        case .success:
+          XCTFail("This should fail")
+        case let .failure(error):
+          XCTAssertEqual(error.pubNubError, PubNubError(.invalidArguments))
+        }
+        expectation.fulfill()
+      }
+
+    wait(for: [expectation], timeout: 1.0)
+  }
+
+  func testMessageCounts_Error_ServiceNotEnabled() {
+    let expectation = self.expectation(description: "Message Counts Response Received")
+
+    guard let sessions = try? MockURLSession.mockSession(for: ["message_counts_error_serviceNotEnabled"]) else {
+      return XCTFail("Could not create mock url session")
+    }
+
+    PubNub(configuration: config, session: sessions.session)
+      .messageCounts(channels: testSingleChannel, timetoken: 0) { result in
+        switch result {
+        case .success:
+          XCTFail("This should fail")
+        case let .failure(error):
+          XCTAssertEqual(error.pubNubError, PubNubError(.messageHistoryNotEnabled))
+        }
+        expectation.fulfill()
+      }
+
+    wait(for: [expectation], timeout: 1.0)
+  }
+}
+
+// MARK: - MessageHistory ResponseDecoder
+
+extension HistoryRouterTests {
+
+  func testFetchV2_decode_jsonDataDecodingFailure() {
+    let decoder = MessageHistoryResponseDecoder()
+    let router = HistoryRouter(.fetchV2(channel: testChannel, max: nil, start: nil, end: nil, includeMeta: false), configuration: config)
+    let response = EndpointResponse(router: router,
+                                    request: .init(url: URL(string: "http://example.com")!),
+                                    response: .init(),
+                                    payload: Data())
+
+    switch decoder.decode(response: response) {
+    case .success:
+      XCTFail("Decoder should not succeed")
+    case .failure(let error):
+      XCTAssertEqual(error.pubNubError?.reason, .jsonDataDecodingFailure)
+    }
+  }
+
+  func testFetchV2_decode_malformedResponse() {
+    guard let data = AnyJSON([]).jsonData, let url = URL(string: "http://example.com") else {
+      return XCTFail("Could not create JSON data")
+    }
+    let decoder = MessageHistoryResponseDecoder()
+    let router = HistoryRouter(.fetchV2(channel: testChannel, max: nil, start: nil, end: nil, includeMeta: false), configuration: config)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Line should be 120 characters or less: currently 137 characters](https://app.codacy.com/app/PubNub/swift/pullRequest?prid=4513555)